### DISCUSSION TITLE
Rendering artifacts

### DISCUSSION THREAD

**Discussion by Nordsoft91** [axmolengine/axmol#1608] [Q&A] [answered]
While rendering of tiles I'm observing graphical artifacts:
<img width="124" alt="Снимок экрана 2023-12-28 в 17 37 48" src="https://github.com/axmolengine/axmol/assets/9308612/6454a5ad-125e-4ba0-9ba1-33acc7b08188">

I use tiles 32x32 pix and place them one by one as Sprite instances in same layer in the same z-order.

UPD:
- MacOS M1
- axmol 2.1.0-rc7

Steps to reproduce:
1. Copy image attached to content folder (or use any image with alpha transparency, but on first example above I used image without alpha channel)
![grid_color](https://github.com/axmolengine/axmol/assets/9308612/14d14e5f-2d45-4e3e-af4d-a8c70fca039d)
2.. add following code into scene::init
```c++
for(int j = 0; j < 20; ++j)
{
	for(int i = 0; i < 20; ++i)
	{
		auto * s = ax::Sprite::create("grid_color.png", ax::Rect(192, 0, 32, 32));
		s->setPosition(i * 32, j * 32);
		addChild(s);
	}
}
```

Expected result: seamless colored rectangle
Actual results: grid appeared
<img width="504" alt="Снимок экрана 2023-12-30 в 23 44 52" src="https://github.com/axmolengine/axmol/assets/9308612/42d2c074-4309-445d-85e8-b94bd5159643">

**Comment by aismann** [comment]
> Is there way to fix them?

Source example which is working for debug can help to fix.
Only an idea:
Can you try to build with axmol 2.0.0,  same behavior here?

**Comment by rh101** [comment]
> While rendering of tiles I'm observing graphical artifacts: <img alt="Снимок экрана 2023-12-28 в 17 37 48" width="124" src="https://private-user-images.githubusercontent.com/9308612/293242945-6454a5ad-125e-4ba0-9ba1-33acc7b08188.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTIiLCJleHAiOjE3MDM4MDQ3MDMsIm5iZiI6MTcwMzgwNDQwMywicGF0aCI6Ii85MzA4NjEyLzI5MzI0Mjk0NS02NDU0YTVhZC0xMjVlLTRiYTAtOWJhMS0zM2FjYzdiMDgxODgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDIzMTIyOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMjhUMjMwMDAzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZGQzODc2ZGI4MDFmNTQxNTllYTg3NzBhZGRjMTc2ZDgxM2ZiMWE4MGJhNzQ5ODM5MWMxNGQyNjFjNGFjNDVhZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.0Mmu7DzSa_tM6TBjsRTBbUyNM4J9Sk41m9T2Qga1nIQ">
> 
> I use tiles 32x32 pix and place them one by one as Sprite instances in same layer in the same z-order.
> 
> Is there way to fix them?

Is there any particular reason the bug report template wasn't used? Refer to this:
https://github.com/axmolengine/axmol/issues/new/choose

It would have prompted you to add information that is missing in your post. 

What you've written provides no useful information in order to even attempt to help you with this.  

For example:

- Which operating system, which platform etc. etc.
- Show the exact source images you're using that cause this issue 
- Show source code that reproduces this issue.  It's not obvious if you're referring to the `TileMap` support, or manually placing images that you're calling "tiles".
- Show what you're actually seeing (as you already have), and what you expect to see.

**Comment by Nordsoft91** [comment]
@rh101 , updated

**Comment by rh101** [comment]
@Nordsoft91

Option 1: 
In `core/base/Config.h`, set `AX_FIX_ARTIFACTS_BY_STRECHING_TEXEL` to 1, and see if that helps.

Option 2: 
The better way to fix this would be to either use a sprite sheet with sufficient padding between the separate frames to avoid the graphical artifacts, or use individual image files. Leave `AX_FIX_ARTIFACTS_BY_STRECHING_TEXEL` as 0.

### Using an individual image:
![grid_color-6](https://github.com/axmolengine/axmol/assets/8603230/73caac5f-e583-4c10-80da-e543d058a6bd)
and this code:
```
for (int j = 0; j < 20; ++j)
{
    for (int i = 0; i < 20; ++i)
    {
        auto* s = ax::Sprite::create("grid_color-6.png");
        s->setPosition(i * 32, j * 32);
        addChild(s);
    }
}
```
Results in this output:
![image](https://github.com/axmolengine/axmol/assets/8603230/9541b161-4c2c-4d26-9461-9bf3550886ce)

### Using a sprite sheet (border = 1, shape padding = 2, extrude = 1 in TexturePacker):
![grid_color_spritesheet](https://github.com/axmolengine/axmol/assets/8603230/9d0cc66d-619b-4fcd-8e21-631f040ad72c)
This archive contains the PNG and PLIST generated by splitting the frames from your original image, then combining them into the correct sprite sheet in TexturePacker: [grid_color_spritesheet.zip](https://github.com/axmolengine/axmol/files/13799537/grid_color_spritesheet.zip)

Using this code to load the sprite sheet and generate the sprite from one of the frames:
```
SpriteFrameCache::getInstance()->addSpriteFramesWithFile("grid_color_spritesheet.plist");

for (int j = 0; j < 20; ++j)
{
    for (int i = 0; i < 20; ++i)
    {
        auto* s = ax::Sprite::createWithSpriteFrameName("grid_color-6.png");
        s->setPosition(i * 32, j * 32);
        addChild(s);
    }
}
```

Results in this:
![image](https://github.com/axmolengine/axmol/assets/8603230/0f5b6f13-6b76-4cc5-af89-77fb1b2ed592)

More info regarding the texture packer settings and why they are required: 
https://www.codeandweb.com/texturepacker/documentation/texture-settings#shape-padding---shape-padding-number
https://www.codeandweb.com/texturepacker/documentation/texture-settings#extrude---extrude-number

**Comment by aismann** [comment]
@rh101 
Good explanation, thanks!

@halx99 
I mean:  It should be part of the documentation

**Comment by Nordsoft91** [comment, marked as answer]
confirm fix with AX_FIX_ARTIFACTS_BY_STRECHING_TEXEL = 1, will redo all textures with TexturePacker later, thanks!

**Comment by rh101** [comment]
> confirm fix with AX_FIX_ARTIFACTS_BY_STRECHING_TEXEL = 1, will redo all textures with TexturePacker later, thanks!

If this is all sorted for you, can you please close this issue?

**Comment by aismann** [comment]
@Nordsoft91 , @halx99 
Please close it

**Comment by aismann** [comment]
I added a wiki page:
https://github.com/axmolengine/axmol/wiki/Renedering-artefacts-with-'tile'-based-textures
